Mathlib/Algebra/Free.lean

```diff
@@ -621,198 +582,183 @@ variable {β : Type u}
+
+/-- `FreeSemigroup` is traversable. -/
+@[to_additive "`FreeAddSemigroup` is traversable."]
+-- Porting note: added noncomputable
+protected noncomputable def traverse {m : Type u → Type u} [Applicative m] {α β : Type u}
+    (F : α → m β) (x : FreeSemigroup α) : m (FreeSemigroup β) :=
+  recOnPure x (fun x ↦ pure <$> F x) fun _x _y ihx ihy ↦ (· * ·) <$> ihx <*> ihy
 #align free_semigroup.traverse FreeSemigroup.traverse
 
 @[to_additive]
-instance : Traversable FreeSemigroup :=
-  ⟨@FreeSemigroup.traverse⟩
+-- Porting note: added noncomputable
+noncomputable instance : Traversable FreeSemigroup := ⟨@FreeSemigroup.traverse⟩
 
 variable {m : Type u → Type u} [Applicative m] (F : α → m β)
 
-@[simp, to_additive]
-theorem traverse_pure (x) : traverse F (pure x : FreeSemigroup α) = pure <$> F x :=
-  rfl
+@[to_additive (attr := simp)]
+theorem traverse_pure (x) : traverse F (pure x : FreeSemigroup α) = pure <$> F x := rfl
 #align free_semigroup.traverse_pure FreeSemigroup.traverse_pure
 
-@[simp, to_additive]
-theorem traverse_pure' : traverse F ∘ pure = fun x => (pure <$> F x : m (FreeSemigroup β)) :=
-  rfl
+@[to_additive (attr := simp)]
+theorem traverse_pure' : traverse F ∘ pure = fun x ↦ (pure <$> F x : m (FreeSemigroup β)) := rfl
 #align free_semigroup.traverse_pure' FreeSemigroup.traverse_pure'
 
 section
 
 variable [LawfulApplicative m]
 
-@[simp, to_additive]
+@[to_additive (attr := simp)]
 theorem traverse_mul (x y : FreeSemigroup α) :
     traverse F (x * y) = (· * ·) <$> traverse F x <*> traverse F y :=
   let ⟨x, L1⟩ := x
   let ⟨y, L2⟩ := y
-  List.recOn L1 (fun x => rfl)
-    (fun hd tl ih x =>
-      show
+  List.recOn L1 (fun x ↦ rfl)
+    (fun hd tl ih x ↦ show
         (· * ·) <$> pure <$> F x <*> traverse F (mk hd tl * mk y L2) =
           (· * ·) <$> ((· * ·) <$> pure <$> F x <*> traverse F (mk hd tl)) <*> traverse F (mk y L2)
-        by rw [ih] <;> simp only [(· ∘ ·), (mul_assoc _ _ _).symm, functor_norm])
+        by rw [ih] ; simp only [(· ∘ ·), (mul_assoc _ _ _).symm, functor_norm])
     x
 #align free_semigroup.traverse_mul FreeSemigroup.traverse_mul
 
-@[simp, to_additive]
+@[to_additive (attr := simp)]
 theorem traverse_mul' :
-    Function.comp (traverse F) ∘ @Mul.mul (FreeSemigroup α) _ = fun x y =>
-      (· * ·) <$> traverse F x <*> traverse F y :=
-  funext fun x => funext fun y => traverse_mul F x y
+    Function.comp (traverse F) ∘ @Mul.mul (FreeSemigroup α) _ = fun x y ↦
+      (· * ·) <$> traverse F x <*> traverse F y := funext fun x ↦ funext fun y ↦ traverse_mul F x y
 #align free_semigroup.traverse_mul' FreeSemigroup.traverse_mul'
 
 end
 
-@[simp, to_additive]
-theorem traverse_eq (x) : FreeSemigroup.traverse F x = traverse F x :=
-  rfl
+@[to_additive (attr := simp)]
+theorem traverse_eq (x) : FreeSemigroup.traverse F x = traverse F x := rfl
 #align free_semigroup.traverse_eq FreeSemigroup.traverse_eq
 
-@[simp, to_additive]
+-- Porting note: deleted simp attribute (simpNF lint)
```

Comment:
What is the linters complaint?

Reply:
I have fixed the `simpNF` linter complains according to the wiki. In fact, in all cases, they are proved by `rfl` and `dsimp` cannot prove the result so I put back the `simp`. 

I do still have a problem with 
```lean
@[to_additive (attr := simp, nolint simpNF)]
theorem mul_eq (x y : FreeMagma α) : mul x y = x * y := rfl
```
The linter still complains about lemmas that I cannot find 
```
@FreeAddMagma.add.sizeOf_spec /- Left-hand side simplifies from
  sizeOf (FreeAddMagma.add a✝¹ a✝)
to
  sizeOf (a✝¹ + a✝)
using
  simp only [@FreeAddMagma.add_eq]
Try to change the left-hand side to the simplified term!
 -/
#check @FreeAddMagma.add.injEq /- Left-hand side simplifies from
  FreeAddMagma.add a✝³ a✝² = FreeAddMagma.add a✝¹ a✝
to
  a✝³ + a✝² = a✝¹ + a✝
using
  simp only [@FreeAddMagma.add_eq]
Try to change the left-hand side to the simplified term!
 -/
#check @FreeMagma.mul.injEq /- Left-hand side simplifies from
  FreeMagma.mul a✝³ a✝² = FreeMagma.mul a✝¹ a✝
to
  a✝³ * a✝² = a✝¹ * a✝
using
  simp only [@FreeMagma.mul_eq]
Try to change the left-hand side to the simplified term!
 -/
#check @FreeMagma.mul.sizeOf_spec /- Left-hand side simplifies from
  sizeOf (FreeMagma.mul a✝¹ a✝)
to
  sizeOf (a✝¹ * a✝)
using
  simp only [@FreeMagma.mul_eq]
Try to change the left-hand side to the simplified term!
 -/

Reply:
These seem to be autogenerated by the inductive definition, so we cannot modify them. The most meaningful thing we can do here is add a nolint attribute manually to them, since I don't think we want to remove the mul_eq lemma. I made that change.